datadog_checks_dev/datadog_checks/dev/tooling/commands/validate/licenses.py

```diff
@@ -310,35 +306,42 @@ def parse_license_path(tar_file_name):
     return None
 
 
-def generate_tarfiles(tar_path):
-    """
-    Streams tarball archive and generates tarfile for each file.
-    """
-    stream = requests.get(tar_path, stream=True)
-    with tarfile.open(fileobj=stream.raw, mode="r|gz") as tar_file:
-        for tar_info in tar_file:
-            if not tar_info.islnk() and not tar_info.issym():
-                yield tar_info, tar_file
-    stream.close()
+def generate_from_tarball(stream):
+    with tarfile.open(fileobj=stream, mode="r|gz") as archive:
+        for tar_info in archive:
+            if not (tar_info.islnk() or tar_info.issym()) and parse_license_path(tar_info.name):
+                fh = archive.extractfile(tar_info)
+                if fh is not None:
+                    with closing(fh) as safe_fh:
+                        yield safe_fh.read()
+
 
+def generate_from_wheel(stream):
+    with ZipFile((stream)) as archive:
+        for name in archive.namelist():
+            if parse_license_path(name):
+                with archive.open(name) as fh:
+                    yield fh.read()
 
-def scrape_copyright_data(url_path, ctx):
+
+def pick_file_generator(url):
+    if url.endswith(".tar.gz"):
+        return generate_from_tarball
+    elif url.endswith(".whl"):
+        return generate_from_wheel
+    else:
+        raise ValueError(f"Unknown type of archive based on this url: {url}")
+
+
+def scrape_copyright_data(url_path):
     """
     Scrapes each tarfile for copyright attributions.
     """
-    tar_path, created_date = probe_github(url_path, ctx)
-    if tar_path:
-        for tar_info, tar_file in generate_tarfiles(tar_path):
-            local_path = parse_license_path(tar_info.name)
-            if local_path:
-                file = tar_file.extractfile(tar_info)
-                cp = find_cpy(file.read())
-                if cp:
-                    file.close()
-                    return cp, created_date
-                if file:
-                    file.close()
-    return None, created_date
+    with closing(requests.get(url_path, stream=True)) as resp:
+        for fcontents in pick_file_generator(url_path)(io.BytesIO(resp.content)):
```

Comment:
`resp.content` triggers a full read into memory thereby removing the benefit of `stream=True`, is there another way? Though it might be the case that incremental reading does not work for wheels/zip files

Reply:
I think this is a good idea:

> One workaround is to keep the streaming setup for tarballs and give up on it for ZipFiles, but prefer tarball pypi URLs instead of grabbing just the first one.

Reply:
Uhhh, I mixed something up: https://urllib3.readthedocs.io/en/latest/reference/urllib3.response.html#urllib3.response.HTTPResponse.seek